### ISSUE TITLE
Temporary directories do not get deleted

### ISSUE THREAD

**Issue by cdiener** [jotech/gapseq#68]
Hi,

I noticed that `gapseq_find.sh` never cleans up the temporary directories it creates. After some parallel runs this will leave behind *a lot of* directories in `/tmp`. Would be nice if the script cloud could clean those up. :sweat_smile:

**Comment by jotech** [comment]
Hi, yes good point! In the past, it was quite useful to have the temp folders for debugging or reusing. By default the folders are deleted by the operation system either on startup and/or via cronjob on a daily schedule.
Hm but I agree that many parallel tasks can lead to larger burden on /tmp. I will integrate the removal on exit!

**Comment by jotech** [comment]
please re-open if there are any issues left

**Comment by cdiener** [comment]
I still see many temp files generated that are not cleaned up. Those are all files, not directories, stored in the top /tmp directory. Those contain content such as this:

```
>sp|P9WQP3|A85A_MYCTU Diacylglycerol acyltransferase/mycolyltransferase Ag85A OS=Mycobacterium tuberculosis (strain ATCC 25618 / H37Rv) OX=83332 GN=fbpA PE=1 SV=1
MQLVDRVRGAVTGMSRRLVVGAVGAALVSGLVGAVGGTATAGAFSRPGLPVEYLQVPSPS
MGRDIKVQFQSGGANSPALYLLDGLRAQDDFSGWDINTPAFEWYDQSGLSVVMPVGGQSS
FYSDWYQPACGKAGCQTYKWETFLTSELPGWLQANRHVKPTGSAVVGLSMAASSALTLAI
YHPQQFVYAGAMSGLLDPSQAMGPTLIGLAMGDAGGYKASDMWGPKEDPAWQRNDPLLNV
GKLIANNTRVWVYCGNGKPSDLGGNNLPAKFLEGFVRTSNIKFQDAYNAGGGHNGVFDFP
DSGTHSWEYWGAQLNAMKPDLQRALGATPNTGPAPQGA
>sp|P9WQP1|A85B_MYCTU Diacylglycerol acyltransferase/mycolyltransferase Ag85B OS=Mycobacterium tuberculosis (strain ATCC 25618 / H37Rv) OX=83332 GN=fbpB PE=1 SV=1
MTDVSRKIRAWGRRLMIGTAAAVVLPGLVGLAGGAATAGAFSRPGLPVEYLQVPSPSMGR
DIKVQFQSGGNNSPAVYLLDGLRAQDDYNGWDINTPAFEWYYQSGLSIVMPVGGQSSFYS
DWYSPACGKAGCQTYKWETFLTSELPQWLSANRAVKPTGSAAIGLSMAGSSAMILAAYHP
QQFIYAGSLSALLDPSQGMGPSLIGLAMGDAGGYKAADMWGPSSDPAWERNDPTQQIPKL
VANNTRLWVYCGNGTPNELGGANIPAEFLENFVRSSNLKFQDAYNAAGGHNAVFNFPPNG
THSWEYWGAQLNAMKGDLQSSLGAG
>sp|P9WQN9|A85C_MYCTU Diacylglycerol acyltransferase/mycolyltransferase Ag85C OS=Mycobacterium tuberculosis (strain ATCC 25618 / H37Rv) OX=83332 GN=fbpC PE=1 SV=1
MTFFEQVRRLRSAATTLPRRLAIAAMGAVLVYGLVGTFGGPATAGAFSRPGLPVEYLQVP
SASMGRDIKVQFQGGGPHAVYLLDGLRAQDDYNGWDINTPAFEEYYQSGLSVIMPVGGQS
SFYTDWYQPSQSNGQNYTYKWETFLTREMPAWLQANKGVSPTGNAAVGLSMSGGSALILA
AYYPQQFPYAASLSGFLNPSEGWWPTLIGLAMNDSGGYNANSMWGPSSDPAWKRNDPMVQ
IPRLVANNTRIWVYCGNGTPSDLGGDNIPAKFLEGLTLRTNQTFRDTYAADGGRNGVFNF
PPNGTHSWPYWNEQLVAMKADIQHVLNGATPPAAPAAPAA
/tmp/tmp.aZnngaGOVm (END)
```

And those are several hundred files per gapseq run.

**Comment by Waschina** [comment]
Hi @cdiener 

thank you for reporting this! I identified the lines in `src/gapseq_find.sh`, that generate and use these temporary files, which do not get deleted after the run. 

Changes that should fix the issue are currently on the branch [tmpfiles_handling](https://github.com/jotech/gapseq/tree/tmpfiles_handling). We will run some tests to make sure the changes do not break something else. When done I'll merge the changes.

**Comment by cdiener** [comment]
Awesome, thanks for looking into this.

**Comment by cdiener** [comment]
Interesting that this seems to come from downloading sequences because this was run with [enforced offline mode](https://github.com/Gibbons-Lab/pipelines/blob/8dc6e00337cfd62b4ffaa71c741e51068746f339/model_builder/main.nf#L131) So I would have thought it would not try to download additional sequences. Though, looking at the code it looks like it still tries to download sequences even with the `-O` option, so maybe this is by design.

**Comment by Waschina** [comment]
I don't think those temporary files that you noticed and which were not cleaned up were from coming from downloading sequences. 

Or did you see in the logs, that it was attempting to download sequences even with the `-O` option?

**Comment by cdiener** [comment]
It's definitely trying to though it can't since that is in a docker container where it does not have write permissions:

```
/opt/gapseq/src/uniprot.sh: line 95: ../updating.log: Read-only file system
touch: cannot touch 'P00367.fasta': Read-only file system
md5sum: P00367.fasta: No such file or directory
cat: P00367.fasta: No such file or directory
P00367.fasta: Read-only file system
md5sum: P00367.fasta: No such file or directory
cat: P00367.fasta: No such file or directory
/opt/gapseq/src/uniprot.sh: line 95: ../updating.log: Read-only file system
touch: cannot touch 'P49448.fasta': Read-only file system
md5sum: P49448.fasta: No such file or directory
cat: P49448.fasta: No such file or directory
P49448.fasta: Read-only file system
md5sum: P49448.fasta: No such file or directory
cat: P49448.fasta: No such file or directory
/opt/gapseq/src/uniprot.sh: line 95: ../updating.log: Read-only file system
touch: cannot touch 'P93541.fasta': Read-only file system
md5sum: P93541.fasta: No such file or directory
cat: P93541.fasta: No such file or directory
P93541.fasta: Read-only file system
md5sum: P93541.fasta: No such file or directory
cat: P93541.fasta: No such file or directory
touch: cannot touch '/opt/gapseq/src/../dat/seq/Bacteria/rxn/GLUTAMATE-DEHYDROGENASE-NADP+-RXN.fasta': Read-only file system
...
```

Looking at [here](https://github.com/jotech/gapseq/blob/a31b45cdde95f21f85e72b32cd3fea37915098a1/src/gapseq_find.sh#L545) I could not find any check for the offline mode before attempting to download.

**Comment by Waschina** [comment]
No that good to know. That Download bug needs a fix, too :)

**Comment by Waschina** [comment]
The fix for the temporary files is now also part of the master branch, along with the the fix of issue #139 .

Please re-open the issue if issue remains or occurs again. Thanks again!
